code/missionui/redalert.cpp

```diff
@@ -776,23 +778,25 @@ void red_alert_store_wingman_status()
 }
 
 // Delete a ship in a red alert mission (since it must have died/departed in the previous mission)
-void red_alert_delete_ship(ship *shipp, int ship_state)
+void red_alert_delete_ship(int shipnum, int ship_state)
 {
-	if ( (shipp->wing_status_wing_index >= 0) && (shipp->wing_status_wing_pos >= 0) ) {
-		if (ship_state == RED_ALERT_DESTROYED_SHIP_CLASS) {
-		hud_set_wingman_status_dead(shipp->wing_status_wing_index, shipp->wing_status_wing_pos);
-		} else if (ship_state == RED_ALERT_PLAYER_DEL_SHIP_CLASS) {
-			hud_set_wingman_status_none(shipp->wing_status_wing_index, shipp->wing_status_wing_pos);
-		} else {
-			Error(LOCATION, "Red Alert: asked to delete ship (%s) with invalid ship state (%d)", shipp->ship_name, ship_state);
-		}
-	}
+	ship* shipp = &Ships[shipnum];
+	int cleanup_mode;	// See ship.h for cleanup mode defines. SHIP_VANISHED etc.
 
-	ship_add_exited_ship( shipp, Ship::Exit_Flags::Player_deleted );
-	obj_delete(shipp->objnum);
-	if ( shipp->wingnum >= 0 ) {
-		ship_wing_cleanup( SHIP_INDEX(shipp), &Wings[shipp->wingnum] );
+	switch (ship_state) {
+	case RED_ALERT_DESTROYED_SHIP_CLASS:
+		cleanup_mode = SHIP_DESTROYED_REDALERT;
+		break;
+	case RED_ALERT_PLAYER_DEL_SHIP_CLASS:
+		cleanup_mode = SHIP_DEPARTED_REDALERT;
+		break;
+	default:
+		Assertion(false, "Red-alert: Unknown delete state '%i'\n", ship_state);
+		break;
```

Comment:
We still need to set `cleanup_mode` here, or it'll be uninitialized in a release build if this branch is ever taken.